service/history/visibilitySamplingClient.go

```diff
@@ -0,0 +1,167 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+
+	"github.com/uber-common/bark"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+)
+
+const (
+	// To sample visibility request, open has only 1 bucket, closed has 2
+	numOfPriorityForOpen   = 1
+	numOfPriorityForClosed = 2
+)
+
+type visibilitySamplingClient struct {
+	rateLimitersForOpen   *domainToBucketMap
+	rateLimitersForClosed *domainToBucketMap
+	persistence           persistence.VisibilityManager
+	config                *Config
+	metricClient          metrics.Client
+	logger                bark.Logger
+}
+
+var _ persistence.VisibilityManager = (*visibilitySamplingClient)(nil)
+
+// NewVisibilitySamplingClient creates a client to manage visibility with sampling
+func NewVisibilitySamplingClient(persistence persistence.VisibilityManager, config *Config, metricClient metrics.Client, logger bark.Logger) persistence.VisibilityManager {
+	return &visibilitySamplingClient{
+		persistence:           persistence,
+		rateLimitersForOpen:   newDomainToBucketMap(),
+		rateLimitersForClosed: newDomainToBucketMap(),
+		config:                config,
+		metricClient:          metricClient,
+		logger:                logger,
+	}
+}
+
+type domainToBucketMap struct {
+	sync.RWMutex
+	mappings map[string]common.PriorityTokenBucket
+}
+
+func newDomainToBucketMap() *domainToBucketMap {
+	return &domainToBucketMap{
+		mappings: make(map[string]common.PriorityTokenBucket),
+	}
+}
+
+func (m *domainToBucketMap) getRateLimiter(domain string, numOfPriority, qps int) common.PriorityTokenBucket {
+	m.RLock()
+	rateLimiter, exist := m.mappings[domain]
+	m.RUnlock()
+
+	if exist {
+		return rateLimiter
+	}
+
+	m.Lock()
+	if rateLimiter, ok := m.mappings[domain]; ok { // read again to ensure no duplicate create
+		m.Unlock()
+		return rateLimiter
+	}
+	if numOfPriority == 1 {
+		rateLimiter = common.NewPriorityTokenBucket(numOfPriority, qps, common.NewRealTimeSource())
+	} else {
+		rateLimiter = common.NewFullPriorityTokenBucket(numOfPriority, qps, common.NewRealTimeSource())
```

Comment:
I think we can just simplify this code to not have if-else branch and always use FullPriorityTokenBucket